docs/data/data-grid/server-side-data/tree-data.md

```diff
@@ -2,14 +2,80 @@
 title: React Server-side tree data
 ---
 
-# Data Grid - Server-side tree data [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')🚧
+# Data Grid - Server-side tree data [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')
 
 <p class="description">Tree data lazy-loading with server-side data source.</p>
 
-:::warning
-This feature isn't implemented yet. It's coming.
+To dynamically load tree data from the server, including lazy-loading of children, you must create a data source and pass the `unstable_dataSource` prop to the Data Grid, as detailed in the [overview section](/x/react-data-grid/server-side-data/).
 
-👍 Upvote [issue #3377](https://github.com/mui/mui-x/issues/3377) if you want to see it land faster.
+Additionally, you must supply the following required props, listed and explained below.
 
-Don't hesitate to leave a comment on the same issue to influence what gets built. Especially if you already have a use case for this component, or if you are facing a pain point with the [currently proposed workaround](https://mui.com/x/react-data-grid/tree-data/#children-lazy-loading).
+```tsx
+<DataGridPro
+  {...data}
+  unstable_dataSource={dataSource}
+  getGroupKey={getGroupKey}
+  hasChildren={hasChildren}
```

Comment:
While writing https://github.com/mui/mui-x/pull/12317#discussion_r1607369601, I thought about `hasChildren` and `getChildrenCount` - do we need to have both of them? Can't we only keep `getChildrenCount`, and returning 0 would mean that the row has no children?

Imagine you have an endpoint that returns the number of children for the row id.
For the group rows, you'll have to call it twice – once in `hasChildren` to return a boolean, and once in `getChildrenCount` to return a number.

Are there use cases I'm missing here?

Reply:
Ah, I see. I think there's a small difference in understanding for this part of the implementation. Thanks @cherniavskii for asking the question.

Props `hasChildren`, `getChildrenCount`, `getGroupKey` are by-design not meant to be `Promise`s, rather they are just like `getRowId` prop. The data supplied to the `getRows` method should contain the information which these methods are supposed to extract.

Let's assume this is a response of a `getRows` call:
```js
{
  rows: [
    {
      id: 1,
      name: 'Parent 1',
      descendantsCount: 10,
    },
    {
      id: 2,
      name: 'Parent 2',
      descendantsCount: 5,
    },
    {
      id: 3,
      name: 'Leaf 1',
      descendantsCount: 0,
    },

    {
      id: 4,
      name: 'Leaf 2',
      descendantsCount: 0,
    },
  ],
  rowCount: 4,
}
```

For this case, the props  `hasChildren`, `getChildrenCount`, `getGroupKey` would look like:
```js
function hasChildren(row) {
  return row.descendantsCount > 0;
}

function getChildrenCount(row) {
  return row.descendantsCount;
}

function getGroupKey(row) {
  return row.name;
}
```

So considering this, it shouldn't be needed to do separate calls for `getChildrenCount` and `hasChildren` since this has to be part of row data itself.

## Motivation behind this design choice

### 1. UX

We need to determine whether or not each row has children on server at the time of rows pre-processing for each specific level to be able to show an expand icon (and schedule expansion if that row has to be expanded as provided by user in `defualtGroupingExpansionDepth` or `isGroupExpandedByDefault` props).

If we don't have that information at the time when `getRows` response is received for a certain level and we have to fetch it afterwards, it may cause a weird UX, (i.e no parent rows initially and as soon as the responses are received, specific rows pop-up to be parents)

**For example**

Consider a tree data with 100 rows each level. On root level `getRows` call, i.e. corresponding to the `groupKeys` value `[]`, there will be 100 rows received, at this point, the Data Grid doesn't know which one of them have children, and if so, how much.

Futher, either of the following could happen:

1. The Data Grid initially shows all the rows as `leaf` rows (no expand icon) and calls `props.hasChildren` in the background for each row (i.e. 100 API calls), not to mention 200 more calls if `getGroupKey` and `getChildrenCount` also correspond to some API endpoint. On receiving the responses, the Data Grid keeps updating the rows to be parents and add their descendant counts as we receive the information (and schedule them to be auto fetched if they are warranted to be).
2. The data grid shows loading indicator until the information of all the rows has been fetched, once all the rows have been fetched, pre-process the row tree, and render it to the users.

And this process will repeat for each group expansion.

To me, 1 is definitely a bad UX, even worse on a slow network. For 2, it will also be weird, for there's always possibility of a failed API call, which might make it difficult to render correct results.

### 2. Server-fatigue

As per my experience with servers and design of APIs, more API calls mean more parallel threads on server, more time in network I/0, and a worse user experience. In general, we should try to reduce the API calls as much as possible, if I am not wrong, that was one of the main motivations behind the introduction of GraphQL.

So, if we consider the same example as in the UX section, its 1 vs 101-301 API calls per level of grid expansion, which might not be a great API design choice.

### 3. Proven design

The competitors like **AgGrid** have been [using](https://www.ag-grid.com/react-data-grid/server-side-model-tree-data/#supplying-tree-data) the similar API and I can't spot any complains on this specific area within the community.

Also, from the [draft](https://www.notion.so/mui-org/Data-source-draft-db58b8b749154420a407f231c5528c1d?pvs=4#08c12d2ddfea46f6b83d915c0084b35e) around the discussion of the data source, I understood that these functions would just be extracting information and actually fetching from the server.

### 4. API Flexibility

If there are some exceptional use-cases, the API is quite flexible to handle them.

For example, to achieve use case no 2 mentioned in the UX section, users can handle this in the getRows method themselves.
The `rows` APIs like `setRows` and `updateRows` are still available, so some advanced use-cases if needed.

If I missed some important use cases that would not be possible with this design, I'd love to know about them.

Thanks

Reply:
Ahh, right.
This makes sense to me!

Reply:
Alright, considering this, are you still convinced it makes sense to make `hasChildren`, `getChildrenCount`, `getGroupKey` part of the data source interface, or keep them part of Grid props to be with their friend `getRowId` 😄

Reply:
Not sure if I missed something but what's the reason for keeping both `hasChildren` and `getChildrenCount`? Could we keep only children count for simplicity?

As for if we should keep them in the DS interface, I think we should if those have no use in any other case.

Reply:
The main intention is, sometimes it's expensive on the backend side to calculate the children count, so the backend tends to only send the information that some children exist. (i.e `hasChildren`) That is the reason, getChildrenCount is optional.
Does it make sense?

Reply:
> are you still convinced it makes sense to make hasChildren, getChildrenCount, getGroupKey part of the data source interface

Yes, I think they belong to the data source.

Reply:
> so the backend may only send the information that some children exist. (hasChildren)

In that case, do we show the expansion button, but not the children count?

Reply:
> In that case, do we show the expansion button, but not the children count?

Yes

Reply:
> Yes, I think they belong to the data source.

Very well then, I'll refactor them! 👍

Reply:
Could we use a special value to indicate that there exist an unknown number of children? `-1` would be a good candidate, and we could export it under a semantic name like `CHILDREN_COUNT_NONZERO`. That would remove the need for `hasChildren`.

Reply:
From a technical perspective, this looks fine, as it will be one less prop to pass, but I am thoughtful about the DX perspective.

IMPO, more specific props would be easier to understand rather than having one prop that users have to understand the interface of. Wdyt?

Would you happen to have another benefit of this in mind that I am not able to guess?

Reply:
I'm actually worried about the DX here, I think that having one less API method to think about when adding a data source is easier for devs to grok.

Reply:
Accepting `-1` as an unknown children count makes sense, as it follows the same API as the unknown row count.
Removed the `hasChildren` prop.